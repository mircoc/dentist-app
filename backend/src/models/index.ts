export const partitionKey = "pk";
export const sortKey = "sk";

export interface BaseSchema {
    created: string;
    modified: string;
    version: number;
}
