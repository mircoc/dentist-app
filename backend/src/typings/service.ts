import { DynamoDB } from "../services/dynamodb";

export interface GenericService {
    close?(): Promise<void>;
}

export interface BaseServicesInstances {
    [key: string]: GenericService | object;
}

export interface ServiceInstances extends BaseServicesInstances {
    dynamoDB: DynamoDB
}

// this declaration must be in scope of the typescript interpreter to work
declare module "fastify" {
    interface FastifyRequest {
        // you must reference the interface and not the type
        services: ServiceInstances;
    }
}