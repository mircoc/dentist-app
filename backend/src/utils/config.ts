import { AppError, AppErrorOptions, ConfigError, ConfigErrorCause } from "./error";

function get(key: string): string {
    const searchKey = key.toUpperCase();
    const val = process.env[searchKey];
    if (val !== undefined) {
        return val;
    }
    throw new ConfigError({ cause: ConfigErrorCause.CONFIG_NOT_FOUND }, { key });
}

export function getConfigString(key: string): string {
    return get(key);
}

export function getConfigNumber(key: string): number {
    const value = Number(get(key));
    if (Number.isNaN(value)) {
        throw new ConfigError({ cause: ConfigErrorCause.CONFIG_NOT_NUMBER }, { key });
    }
    return value;
}