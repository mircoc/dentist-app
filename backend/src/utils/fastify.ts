import fastify, { FastifyError, FastifyInstance, FastifyServerOptions } from "fastify";
import { AppError, NotFoundError, NotFoundErrorCause, ValidationError } from "./error";
import { BaseServicesInstances } from "../typings/service";
import schema from "../schema.json";
import { FastifySchemaValidationError } from "fastify/types/schema";

export async function init<T extends BaseServicesInstances>(
    options: FastifyServerOptions,
    services: T,
): Promise<FastifyInstance> {
    const app = fastify({
        ...options,
        schemaErrorFormatter: (errors: FastifySchemaValidationError[], dataVar: string): Error => {
            return new ValidationError(errors, dataVar);
        },
        ajv: {
            customOptions: { jsonPointers: true, removeAdditional: true },
        },
    });

    app.setNotFoundHandler((request, reply) => {
        const method = request.raw.method || "unknown";
        const url = request.raw.url || "unknown";
        const notFoundError = new NotFoundError({ cause: NotFoundErrorCause.ROUTE_NOT_FOUND }, {});
        const error = {
            code: notFoundError.code,
            details: {
                cause: notFoundError.details.cause,
            },
        };
        app.log.warn(error, `route not found: ${method} ${url}`);
        reply.code(notFoundError.statusCode).send(error);
    });
    app.setErrorHandler((error: AppError | FastifyError | Error, request, reply) => {
        // uncomment this on test execution to show errors
        // console.error("errorHandler: ", error);

        if (error instanceof AppError) {
            app.log.warn(error, `error-handler for AppError: ${error.message}`);

            if (error.headers) {
                reply.headers(error.headers);
            }
            reply.code(error.statusCode || 500).send({
                code: error.code,
                details: error.details,
            });
        } else {
            app.log.error(error, `internal error: ${error.message}`);
            reply.code(500).send({
                code: "INTERNAL_ERROR",
                details: {
                    cause: "INTERNAL_ERROR",
                },
            });
        }
    });

    app.addHook("onClose", async () => {
        for (const serviceName of Object.keys(services)) {
            const service = services[serviceName];
            if (service && "close" in service && service.close) {
                await service.close();
            }
        }
    });
    // save services instances for using inside routes
    app.decorateRequest("services", { getter: () => services }, );

    // load generated json schema
    app.addSchema(schema);

    return app;
}

export interface FastifyStartOptions {
    host: string;
    port: number;
    server: FastifyInstance;
}
export async function start(options: FastifyStartOptions): Promise<void> {
    const { host, port, server } = options;
    try {
        await server.listen(port, host);
        server.log.info(`fastify server ready for requests on ${host}:${port}`);
    } catch (err) {
        server.log.error(err, `caught fastify start error: ${err.message}`);
        process.exit(1);
    }
}
