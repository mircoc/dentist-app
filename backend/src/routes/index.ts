import { FastifyInstance } from "fastify";
import admin from "./admin";
import user from "./user";

export default async function(server: FastifyInstance): Promise<void> {
    server.register(admin, { prefix: "/admin" });
    server.register(user, { prefix: "/user" });
}