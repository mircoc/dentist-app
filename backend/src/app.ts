import { FastifyInstance } from "fastify";
import fastifySwagger from "fastify-swagger";
import P from "pino";
import routes from "./routes";
import { swaggerConfiguration } from "./routes/doc";
import { DynamoDB, DynamoDBOptions } from "./services/dynamodb";
import { ServiceInstances } from "./typings/service";
import { init, start } from "./utils/fastify";
import { getLogger } from "./utils/logger";


export interface AppConfig {
    http: {
        port: number;
        host: string;
    };
    logger: P.Logger;
    dynamoDB: DynamoDBOptions;
}

export async function startApp(appConfig: AppConfig): Promise<FastifyInstance> {
    // FIXME: init dynamodb service
    const dynamoDB = new DynamoDB(
        getLogger("dynamoDb"),
        {
            region: appConfig.dynamoDB.region,
            endpoint: appConfig.dynamoDB.endpoint,
            tableName: appConfig.dynamoDB.tableName,
        },
    )
    debugger;
    // FIXME: define services
    const services: ServiceInstances = {
        dynamoDB,
    };

    // init fastify
    const server = await init(
        {
            logger: appConfig.logger,
        },
        services,
    );

    // add documentation route for the following routes
    server.register(fastifySwagger, swaggerConfiguration);

    // add routes
    server.register(routes, { prefix: "/api/v1" });

    // start server
    await start({
        server,

        ...appConfig.http,
    });

    await server.ready();
    server.swagger();

    return server;
}
