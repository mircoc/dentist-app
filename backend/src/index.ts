import { FastifyInstance } from "fastify";
import { startApp } from "./app";
import { getConfigNumber, getConfigString } from "./utils/config";
import { installFatalHandlers } from "./utils/fatal";
import { getLogger } from "./utils/logger";

const logger = getLogger("index");

installFatalHandlers(logger);


(async () => startApp({
    http: {
        port: getConfigNumber("HTTP_PORT"),
        host: getConfigString("HTTP_HOST"),
    },
    logger,
    dynamoDB: {
        region: getConfigString("DYNAMODB_REGION"),
        endpoint: getConfigString("DYNAMODB_ENDPOINT"),
        tableName: `${getConfigString("DYNAMODB_PREFIX_TABLE")}dentist`,
        jwtSecret: getConfigString("JWT_SECRET"),
    },
}))().catch((error) => {
    logger.fatal(error, `caught error: ${error}`);
});
