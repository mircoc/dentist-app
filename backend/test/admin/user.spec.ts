import { FastifyInstance } from "fastify";
import { startApp } from "../../src/app";
import pino from "pino";
import supertest from "supertest";
import { dynamoEndpoint, dynamoRegion, testTableName } from "../config";
import { clearTable, createTable } from "../table";
import { ServiceInstances } from "../../src/typings/service";

jest.setTimeout(120_000);

// prefix to allow jest parallel executions
const TEST_PREFIX = "admin-user";

const tableName = `${TEST_PREFIX}_${testTableName}`;

describe("Admin user handling", () => {
    let app: FastifyInstance;
    let services: ServiceInstances;
    let authToken;
    const logger = pino().child({ module: "Test E2E" });

    beforeAll(async () => {
        try {
            await createTable(tableName, dynamoRegion, dynamoEndpoint);
            const resultStart = await startApp({
                http: {
                    port: 4200,
                    host: "localhost",
                },
                logger,
                dynamoDB: {
                    region: dynamoRegion,
                    endpoint: dynamoEndpoint,
                    tableName,
                    jwtSecret: "testSecret",
                },
            });
            app = resultStart[0];
            services = resultStart[1];
        } catch (err) {
            console.error(err);
        }
    });
    afterAll(async () => {
        app && (await app.close());
    });
    beforeEach(async () => {
        await clearTable(tableName, dynamoRegion, dynamoEndpoint);

        const dynamoDB = services.dynamoDB;
        const admin = await dynamoDB.createUser({
            userName: "myadmin",
            name: "Mirco",
            surname: "Cipriani",
            bornDate: "1978-07-06",
            fiscalCode: "ABCDED12A11A123A",
            telephone: "390012345",
        });
        // await dynamoDB.resetUserPassword(admin, "testPass");
        const { token } = await dynamoDB.updateUserLogin(admin);
        authToken = token;
    });
    test("Successfully create an user", async () => {
        const result = await supertest(app.server)
            .post("/api/v1/admin/user")
            .send({
                userName: "mircocip",
                name: "Mirco",
                surname: "Cipriani",
                bornDate: "1978-07-06",
                fiscalCode: "ABCDEF12A11A123A",
                telephone: "392012345",
            })
            .set("Authorization", `Bearer ${authToken}`)
            .set("Accept", "application/json")
            .expect("Content-Type", /json/)
            .expect(200);

        //console.log("result.body", result.body);
        expect(result.body).toHaveProperty("name", "Mirco");
    });

    test("Prevent duplicated username", async () => {
        const result = await supertest(app.server)
            .post("/api/v1/admin/user")
            .send({
                userName: "mircocip",
                name: "Mirco",
                surname: "Cipriani",
                bornDate: "1978-07-06",
                fiscalCode: "ABCDEF12A11A123A",
                telephone: "392012345",
            })
            .set("Authorization", `Bearer ${authToken}`)
            .set("Accept", "application/json")
            .expect("Content-Type", /json/)
            .expect(200);

        //console.log("result.body", result.body);
        expect(result.body).toHaveProperty("name", "Mirco");

        const result2 = await supertest(app.server)
            .post("/api/v1/admin/user")
            .send({
                userName: "mircocip",
                name: "Pippo",
                surname: "Baudo",
                bornDate: "1938-07-06",
                fiscalCode: "ABCDEF12A11A123A",
                telephone: "392012345",
            })
            .set("Authorization", `Bearer ${authToken}`)
            .set("Accept", "application/json")
            .expect("Content-Type", /json/)
            .expect(409);

        console.log("result.body", result2.body);
        //expect(result2.body).toHaveProperty("name", "Mirco");
    });

    test("Successfully update a user field", async () => {
        const result = await supertest(app.server)
            .post("/api/v1/admin/user")
            .send({
                userName: "mircocip",
                name: "Mirco",
                surname: "Cipriani",
                bornDate: "1978-07-06",
                fiscalCode: "ABCDEF12A11A123A",
                telephone: "392012345",
            })
            .set("Authorization", `Bearer ${authToken}`)
            .set("Accept", "application/json")
            .expect("Content-Type", /json/)
            .expect(200);

        //console.log("result.body", result.body);
        expect(result.body).toHaveProperty("name", "Mirco");

        const result2 = await supertest(app.server)
            .put(`/api/v1/admin/user/mircocip`)
            .send({
                name: "Pippo",
                surname: "Baudo",
                bornDate: "1938-07-06",
                fiscalCode: "XBCDEF12A11A123A",
                telephone: "399012345",
            })
            .set("Authorization", `Bearer ${authToken}`)
            .set("Accept", "application/json")
            .expect("Content-Type", /json/)
            .expect(200);

        // console.log("result.body", result2.body);
        expect(result2.body).toHaveProperty("name", "Pippo");
        expect(result2.body).toHaveProperty("surname", "Baudo");
        expect(result2.body).toHaveProperty("bornDate", "1938-07-06");
        expect(result2.body).toHaveProperty("fiscalCode", "XBCDEF12A11A123A");
        expect(result2.body).toHaveProperty("telephone", "399012345");
    });

    test("Cannot update a not existing user", async () => {
        const result = await supertest(app.server)
            .post("/api/v1/admin/user")
            .send({
                userName: "mircocip",
                name: "Mirco",
                surname: "Cipriani",
                bornDate: "1978-07-06",
                fiscalCode: "ABCDEF12A11A123A",
                telephone: "392012345",
            })
            .set("Authorization", `Bearer ${authToken}`)
            .set("Accept", "application/json")
            .expect("Content-Type", /json/)
            .expect(200);

        //console.log("result.body", result.body);
        expect(result.body).toHaveProperty("name", "Mirco");

        const result2 = await supertest(app.server)
            .put(`/api/v1/admin/user/nonce`)
            .send({
                name: "Pippo",
                surname: "Baudo",
                bornDate: "1938-07-06",
                fiscalCode: "XBCDEF12A11A123A",
                telephone: "399012345",
            })
            .set("Authorization", `Bearer ${authToken}`)
            .set("Accept", "application/json")
            .expect("Content-Type", /json/)
            .expect(404);

        // console.log("result.body", result2.body);
    });

    test("Successfully list created users", async () => {
        const result = await supertest(app.server)
            .post("/api/v1/admin/user")
            .send({
                userName: "mircocip",
                name: "Mirco",
                surname: "Cipriani",
                bornDate: "1978-07-06",
                fiscalCode: "ABCDEF12A11A123A",
                telephone: "392012345",
            })
            .set("Authorization", `Bearer ${authToken}`)
            .set("Accept", "application/json")
            .expect("Content-Type", /json/)
            .expect(200);

        //console.log("result.body", result.body);
        expect(result.body).toHaveProperty("name", "Mirco");

        const result2 = await supertest(app.server)
            .post(`/api/v1/admin/user`)
            .send({
                userName: "pippobaudo",
                name: "Pippo",
                surname: "Baudo",
                bornDate: "1938-07-06",
                fiscalCode: "XBCDEF12A11A123A",
                telephone: "399012345",
            })
            .set("Authorization", `Bearer ${authToken}`)
            .set("Accept", "application/json")
            .expect("Content-Type", /json/)
            .expect(200);

        // console.log("result.body", result2.body);
        expect(result2.body).toHaveProperty("name", "Pippo");
        expect(result2.body).toHaveProperty("surname", "Baudo");
        expect(result2.body).toHaveProperty("bornDate", "1938-07-06");
        expect(result2.body).toHaveProperty("fiscalCode", "XBCDEF12A11A123A");
        expect(result2.body).toHaveProperty("telephone", "399012345");

        const result3 = await supertest(app.server)
            .get(`/api/v1/admin/user`)
            .set("Authorization", `Bearer ${authToken}`)
            .set("Accept", "application/json")
            .expect("Content-Type", /json/)
            .expect(200);

        console.log("result3.body", result3.body);
        expect(result3.body.count).toBe("3");
        expect(result3.body.data).toHaveLength(3);
        expect(result3.body.data).toMatchSnapshot();
    });
});
