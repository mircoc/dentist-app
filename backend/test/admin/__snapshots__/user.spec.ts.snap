// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin user handling Successfully list created users 1`] = `
Array [
  Object {
    "bornDate": "1978-07-06",
    "fiscalCode": "ABCDEF12A11A123A",
    "name": "Mirco",
    "surname": "Cipriani",
    "telephone": "392012345",
    "userName": "mircocip",
  },
  Object {
    "bornDate": "1938-07-06",
    "fiscalCode": "XBCDEF12A11A123A",
    "name": "Pippo",
    "surname": "Baudo",
    "telephone": "399012345",
    "userName": "pippobaudo",
  },
  Object {
    "bornDate": "1978-07-06",
    "fiscalCode": "ABCDED12A11A123A",
    "name": "Mirco",
    "surname": "Cipriani",
    "telephone": "390012345",
    "userName": "myadmin",
  },
]
`;
